File: backend/app/routers/treasury.py
# app/routers/treasury.py
from typing import Annotated, List, Optional
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession
from app.database import get_db
from app.dependencies import get_current_user, require_permissions
from app.models.user import User
from app.models.treasury import TransactionType, TransactionStatus
from app.schemas.treasury import (
    WalletCreate, WalletUpdate, WalletResponse, WalletWithTransactions,
    TreasuryTransactionCreate, TreasuryTransactionResponse,
    TransactionFilter, TransactionApproval, TreasuryReport
)
from app.services.treasury import TreasuryService


router = APIRouter(prefix="/api/treasury", tags=["treasury"])


@router.get("/wallets", response_model=List[WalletResponse])
async def list_wallets(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    """List all organization wallets."""
    service = TreasuryService(db)
    wallets = await service.get_wallets()
    return wallets


@router.post("/wallets", response_model=WalletResponse, status_code=status.HTTP_201_CREATED)
async def create_wallet(
    data: WalletCreate,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permissions(["org.manage_treasury"]))],
):
    """Create a new organization wallet (admin only)."""
    service = TreasuryService(db)
    wallet = await service.create_wallet(
        name=data.name,
        description=data.description,
        is_primary=data.is_primary
    )
    return wallet


@router.get("/wallets/{wallet_id}", response_model=WalletWithTransactions)
async def get_wallet(
    wallet_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
):
    """Get a specific wallet with its transactions."""
    service = TreasuryService(db)
    wallet = await service.get_wallet(wallet_id)
    if not wallet:
        raise HTTPException(status_code=404, detail="Wallet not found")
    
    transactions = await service.get_transactions(wallet_id, limit=limit, offset=offset)
    return WalletWithTransactions(
        id=wallet.id,
        name=wallet.name,
        description=wallet.description,
        balance=wallet.balance,
        is_primary=wallet.is_primary,
        created_at=wallet.created_at,
        updated_at=wallet.updated_at,
        transactions=transactions
    )


@router.put("/wallets/{wallet_id}", response_model=WalletResponse)
async def update_wallet(
    wallet_id: int,
    data: WalletUpdate,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permissions(["org.manage_treasury"]))],
):
    """Update wallet name and description (admin only)."""
    service = TreasuryService(db)
    try:
        wallet = await service.update_wallet(
            wallet_id=wallet_id,
            name=data.name,
            description=data.description
        )
        return wallet
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.delete("/wallets/{wallet_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_wallet(
    wallet_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permissions(["org.manage_treasury"]))],
):
    """Delete an empty wallet (admin only)."""
    service = TreasuryService(db)
    try:
        await service.delete_wallet(wallet_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/wallets/{wallet_id}/transactions", response_model=TreasuryTransactionResponse, status_code=status.HTTP_201_CREATED)
async def create_transaction(
    wallet_id: int,
    data: TreasuryTransactionCreate,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    """Create a new transaction (deposit auto-approved, withdrawal pending)."""
    service = TreasuryService(db)
    try:
        transaction = await service.create_transaction(
            treasury_id=wallet_id,
            user_id=current_user.id,
            data=data
        )
        return transaction
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/wallets/{wallet_id}/transactions", response_model=List[TreasuryTransactionResponse])
async def list_transactions(
    wallet_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
    type: Optional[TransactionType] = None,
    status: Optional[TransactionStatus] = None,
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
):
    """List transactions for a wallet with optional filters."""
    service = TreasuryService(db)
    transactions = await service.get_transactions(
        treasury_id=wallet_id,
        limit=limit,
        offset=offset,
        type_filter=type,
        status_filter=status
    )
    return transactions


@router.get("/pending", response_model=List[TreasuryTransactionResponse])
async def get_pending_transactions(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permissions(["org.manage_treasury"]))],
    wallet_id: Optional[int] = None,
):
    """Get all pending transactions requiring approval (admin only)."""
    service = TreasuryService(db)
    transactions = await service.get_pending_transactions(wallet_id)
    return transactions


@router.post("/transactions/{transaction_id}/approve", response_model=TreasuryTransactionResponse)
async def approve_transaction(
    transaction_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permissions(["org.manage_treasury"]))],
):
    """Approve a pending withdrawal transaction (admin only)."""
    service = TreasuryService(db)
    try:
        transaction = await service.approve_transaction(transaction_id, current_user.id)
        return transaction
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/transactions/{transaction_id}/reject", response_model=TreasuryTransactionResponse)
async def reject_transaction(
    transaction_id: int,
    data: TransactionApproval,
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(require_permissions(["org.manage_treasury"]))],
):
    """Reject a pending withdrawal transaction (admin only)."""
    service = TreasuryService(db)
    try:
        transaction = await service.reject_transaction(transaction_id, current_user.id)
        return transaction
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/wallets/{wallet_id}/reports/summary")
async def get_wallet_report(
    wallet_id: int,
    days: int = Query(30, ge=1, le=365),
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    """Get financial report for a wallet."""
    service = TreasuryService(db)
    report = await service.get_report(wallet_id, days)
    return report

File: frontend/src/stores/treasury.js
// src/stores/treasury.js
import { defineStore } from 'pinia';
import api from '../services/api';

export const useTreasuryStore = defineStore('treasury', {
  state: () => ({
    wallets: [],
    selectedWallet: null,
    transactions: [],
    pendingTransactions: [],
    report: null,
    isLoading: false,
    error: null,
  }),

  getters: {
    primaryWallet: (state) => state.wallets.find(w => w.is_primary) || state.wallets[0],
    totalBalance: (state) => state.wallets.reduce((sum, w) => sum + w.balance, 0),
    pendingCount: (state) => state.pendingTransactions.length,
  },

  actions: {
    async fetchWallets() {
      this.isLoading = true;
      this.error = null;
      try {
        const response = await api.get('/treasury/wallets');
        this.wallets = response.data;
        if (!this.selectedWallet && this.wallets.length > 0) {
          this.selectedWallet = this.primaryWallet;
        }
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to fetch wallets';
        throw err;
      } finally {
        this.isLoading = false;
      }
    },

    async createWallet(walletData) {
      this.isLoading = true;
      this.error = null;
      try {
        const response = await api.post('/treasury/wallets', walletData);
        this.wallets.push(response.data);
        if (response.data.is_primary || this.wallets.length === 1) {
          this.selectedWallet = response.data;
        }
        return response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to create wallet';
        throw err;
      } finally {
        this.isLoading = false;
      }
    },

    async updateWallet(walletId, walletData) { // Renamed and changed params
      this.isLoading = true;
      this.error = null;
      try {
        // If the updated wallet is set as primary, unmark others locally (backend handles db)
        if (walletData.is_primary) {
          this.wallets.forEach(w => {
            if (w.id !== walletId) {
              w.is_primary = false;
            }
          });
        }

        const response = await api.patch(`/treasury/wallets/${walletId}`, walletData); // Changed to patch and pass data object
        const index = this.wallets.findIndex(w => w.id === walletId);
        if (index !== -1) {
          this.wallets[index] = response.data;
        }
        if (this.selectedWallet?.id === walletId) {
          this.selectedWallet = response.data;
        }
        return response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to update wallet';
        throw err;
      } finally {
        this.isLoading = false;
      }
    },

    async deleteWallet(walletId) {
      this.isLoading = true;
      this.error = null;
      try {
        await api.delete(`/treasury/wallets/${walletId}`);
        this.wallets = this.wallets.filter(w => w.id !== walletId);
        if (this.selectedWallet?.id === walletId) {
          this.selectedWallet = this.wallets[0] || null;
        }
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to delete wallet';
        throw err;
      } finally {
        this.isLoading = false;
      }
    },

    async selectWallet(walletId) {
      const wallet = this.wallets.find(w => w.id === walletId);
      if (wallet) {
        this.selectedWallet = wallet;
        await this.fetchTransactions(walletId);
      }
    },

    async fetchTransactions(walletId, filters = {}) {
      this.isLoading = true;
      this.error = null;
      try {
        const params = new URLSearchParams();
        if (filters.type) params.append('type', filters.type);
        if (filters.status) params.append('status', filters.status);
        if (filters.limit) params.append('limit', filters.limit);
        if (filters.offset) params.append('offset', filters.offset);
        
        const response = await api.get(`/treasury/wallets/${walletId}/transactions?${params}`);
        this.transactions = response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to fetch transactions';
        throw err;
      } finally {
        this.isLoading = false;
      }
    },

    async createTransaction(walletId, type, amount, category, description = '') {
      this.isLoading = true;
      this.error = null;
      try {
        const response = await api.post(`/treasury/wallets/${walletId}/transactions`, {
          type,
          amount,
          category,
          description,
        });
        
        // Update local state
        this.transactions.unshift(response.data);
        
        // Update wallet balance if deposit (auto-approved)
        if (type === 'deposit' && response.data.status === 'approved') {
          const wallet = this.wallets.find(w => w.id === walletId);
          if (wallet) {
            wallet.balance += amount;
          }
        }
        
        return response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to create transaction';
        throw err;
      } finally {
        this.isLoading = false;
      }
    },

    async fetchPendingTransactions() {
      this.isLoading = true;
      this.error = null;
      try {
        const response = await api.get('/treasury/pending');
        this.pendingTransactions = response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to fetch pending transactions';
        throw err;
      } finally {
        this.isLoading = false;
      }
    },

    async approveTransaction(transactionId) {
      this.isLoading = true;
      this.error = null;
      try {
        const response = await api.post(`/treasury/transactions/${transactionId}/approve`);
        
        // Remove from pending
        this.pendingTransactions = this.pendingTransactions.filter(t => t.id !== transactionId);
        
        // Update transaction in list if present
        const index = this.transactions.findIndex(t => t.id === transactionId);
        if (index !== -1) {
          this.transactions[index] = response.data;
        }
        
        // Update wallet balance
        const wallet = this.wallets.find(w => w.id === response.data.treasury_id);
        if (wallet && response.data.type === 'withdrawal') {
          wallet.balance -= response.data.amount;
        }
        
        return response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to approve transaction';
        throw err;
      } finally {
        this.isLoading = false;
      }
    },

    async rejectTransaction(transactionId, notes = '') {
      this.isLoading = true;
      this.error = null;
      try {
        const response = await api.post(`/treasury/transactions/${transactionId}/reject`, {
          approved: false,
          notes,
        });
        
        // Remove from pending
        this.pendingTransactions = this.pendingTransactions.filter(t => t.id !== transactionId);
        
        // Update transaction in list if present
        const index = this.transactions.findIndex(t => t.id === transactionId);
        if (index !== -1) {
          this.transactions[index] = response.data;
        }
        
        return response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to reject transaction';
        throw err;
      } finally {
        this.isLoading = false;
      }
    },

    async fetchReport(walletId, days = 30) {
      this.isLoading = true;
      this.error = null;
      try {
        const response = await api.get(`/treasury/wallets/${walletId}/reports/summary?days=${days}`);
        this.report = response.data;
        return response.data;
      } catch (err) {
        this.error = err.response?.data?.detail || 'Failed to fetch report';
        throw err;
      } finally {
        this.isLoading = false;
      }
    },
  },
});
